### ISSUE TITLE
Improve the model

### ISSUE THREAD

**Issue by rdancer** [rdancer/Tetris-2023#3]
## Problem

The model quickly (~ 100 games) learns not to be utterly rubbish, but soon settles to a performance level where during the course of a game, it sometimes clears a few lines, but nevertheless in a short order sparsely fills the board and loses the game. The model currently takes its cue from the Reward.get_reward() method, but using model.predict() is markedly worse than using the best-scored move straight from the reward function.

## Input, output, reward

The model input is the state of the board, the current's piece shape, and sundries; the input shape is a flat (1, 226) tensor. The output is the desired rotation and position of the piece (4 rotations x 10 column positions = 40 possible moves), a (1, 40)-shaped tensor. The reward function looks ahead and computes the state the board would be after performing each of the 40 possible moves, but the model itself only has access to the  current board state.

Working at the level of piece placement (drop piece in column 3 at rotation 2) should give better results than working on the level of individual piece movements (left, left, left, rotate, left, rotate, drop).

## Potential paths to improvement

* Allow the model to maximise the reward across several moves -- add memory?
* Parametrise the Reward.get_reward() function with learnable TenserFlow parameters, so that the reward function itself improves in order to clear lines

![Screen Recording 2023-01-12 at 15 03 29](https://user-images.githubusercontent.com/51028/212102431-7ed12ef7-56ea-414b-a9e6-528c2e06e1f7.gif)

https://github.com/rdancer/Tetris-2023/blob/f42b1328b704baeff7553f1c2ca687df9bf3c787/ml/model.py#L102-L115

https://github.com/rdancer/Tetris-2023/blob/f42b1328b704baeff7553f1c2ca687df9bf3c787/ml/reward.py#L115-L122

**Comment by rdancer** [comment]
The branch for this issue is [`use-real-model`.](https://github.com/rdancer/Tetris-2023/tree/use-real-model)

**Comment by rdancer** [comment]
Rewrote the reward function, which is now very good b77ff587 -- but this only puts in stark relief how meager the model is.

**Comment by rdancer** [comment]
There is now a [Jupiter notebook](https://github.com/rdancer/Tetris-2023/blob/use-real-model/ml/Model%20Experiments.ipynb) which should make experimenting with the model / training easier.

**Comment by rdancer** [comment]
6497085a is much improved: not state-of-the-art yet, but can play and score a few dozen lines.

![Screen Recording 2023-01-19 at 15 04 40](https://user-images.githubusercontent.com/51028/213478470-6d9a89d5-8538-493f-a6e0-d5cc6ef6a1e1.gif)

**Comment by rdancer** [comment]
### Next steps

* Improve the reward function using a **genetic algorithm** -- possibly use TensorFlow trainable variables for the hyperparameters
* Speed up training
    - Eliminate delays and duplicate episodes by registering a **callback** with script.js:createPiece() to run the next prediction, and likewise fire the next tick straight after Control.drop()
    - Record the training sessions for fast replay later
    - Explode each encountered state into variations of all 7 tetrominoes and all board heights -- because we ignore inaccessible holes, adding/removing lines is trivial
    - Procedurally generate all reasonable bumpiness variations -- we wouldn't need game server for training at all if we did that; or we can record the bumpiness actually encountered, for better fit
    - Experiment with `headless=True`
    - Experiment with parallel training
* Implement **reinforcement learning**
* **Improve the model** and maybe make it smaller / larger
* Implement analytics
    - **TensorBoard**
    - **Keep track of how many episodes trained** for, and save/load that number when saving/loading weights
